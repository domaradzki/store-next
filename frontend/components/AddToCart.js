import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import styled from 'styled-components';
import { ShoppingCart } from '@styled-icons/material/ShoppingCart';

import { CURRENT_USER_QUERY } from './User';

const CartButton = styled.button`
  cursor: pointer;
  position: absolute;
  right: 15px;
  bottom: 15px;
  background: var(--blue);
  border: none;
  border-radius: 50px;
  padding: 10px;
`;

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(productId: $id) {
      id
    }
  }
`;

export default function AddToCart({ id }) {
  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
    variables: { id },
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });
  return (
    <CartButton type="button" disabled={loading} onClick={addToCart}>
      <ShoppingCart size="18" color="#fff" title="Koszyk" />
    </CartButton>
  );
}
