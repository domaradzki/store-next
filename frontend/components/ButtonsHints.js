import { ShoppingBag, Person, Favorite } from '@styled-icons/material';
import styled from 'styled-components';
import SignOut from './SignOut';

import { useUser } from './User';

const Container = styled.div`
  position: absolute;
  top: 40px;
  z-index: 1;
  transition: all 0.2s ease 0s;
  left: auto;
  right: 0px;
  border: 1px solid var(--lightBlack);
  border-radius: 3px;
  width: 340px;
  min-height: 130px;
  background: #fff;
  padding: 10px;
`;

const Hello = styled.h3`
  padding: 10px;
  text-align: center;
  color: var(--blue);
`;

const ButtonsContainer = styled.div`
  display: grid;
  width: 100%;
  grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
  grid-gap: 10px;
  margin-bottom: 10px;
`;

const Button = styled.button`
  width: auto;
  background: ${({ color }) => color || 'var(--blue)'};
  color: ${({ color }) => (color ? 'var(--blue)' : 'var(--white)')};
  border: 2px solid var(--blue);
  font-size: 1.5rem;
  font-weight: 600;
  padding: 6px 16px;
  border-radius: 4px;
`;

const Box = styled.div`
  width: 100%;
  display: grid;
  grid-template-columns: 50px 1fr;
  align-items: center;
  border-top: 1px solid var(--lightBlack);
  padding: 5px;
  .box-title {
    a {
      text-decoration: none;
      color: var(--blue);
      position: relative;
      &:before {
        content: '';
        position: absolute;
        width: 100%;
        height: 2px;
        bottom: -2px;
        left: 0;
        right: 0;
        margin: 0 auto;
        background-color: var(--blue);
        visibility: hidden;
        transform: scaleX(0);
        transition: all 0.2s ease-in-out 0s;
      }
      &:hover:before {
        visibility: visible;
        transform: scaleX(1);
      }
    }
  }
`;

export default function ButtonsHints({
  className,
  onMouseEnter,
  onFocus,
  onMouseLeave,
  setModule,
}) {
  const user = useUser();

  const handleLogin = (e) => {
    setModule(e.target.name);
  };
  return (
    <Container
      className={className}
      onMouseEnter={onMouseEnter}
      onFocus={onFocus}
      onMouseLeave={onMouseLeave}
    >
      {user ? (
        <>
          <Hello>Dzień dobry, {user.name}</Hello>
          <ButtonsContainer>
            <SignOut />
          </ButtonsContainer>
          <Box>
            <Person size="24" color="var(--blue)" title="Profil" />
            <div className="box-title">
              <a>Profil</a>
            </div>
          </Box>
          <Box>
            <Favorite size="24" color="var(--blue)" title="Profil" />
            <div className="box-title">
              <a>Ulubione</a>
            </div>
          </Box>
          <Box>
            <ShoppingBag size="24" color="var(--blue)" title="Profil" />
            <div className="box-title">
              <a>Zamówienia</a>
            </div>
          </Box>
        </>
      ) : (
        <>
          <Hello>Dzień dobry, Piękności !!!</Hello>
          <ButtonsContainer>
            <Button name="login" onClick={handleLogin}>
              Zaloguj się
            </Button>
            <Button name="register" onClick={handleLogin} color="var(--white)">
              Rejestracja
            </Button>
          </ButtonsContainer>
        </>
      )}
    </Container>
  );
}
