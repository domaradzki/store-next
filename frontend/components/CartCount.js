import { CSSTransition, TransitionGroup } from 'react-transition-group';
import styled from 'styled-components';

const Dot = styled.div`
  background: var(--blue);
  color: var(--white);
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 1.2rem;
  min-width: 2rem;
  margin-left: 1.2rem;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
  user-select: none;
`;

const AnimationStyles = styled.span`
  position: absolute;
  top: 12px;
  left: 12px;
  .count {
    display: block;
    position: relative;
    transition: transform 0.4s;
    backface-visibility: hidden;
  }
  .count-enter {
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .cont-exit-active {
    transform: scale(4) rotateX(0.5turn);
    background: var(--dark);
  }
`;

export default function CartCount({ count }) {
  return (
    <AnimationStyles>
      <TransitionGroup>
        <CSSTransition
          unmontOnExit
          className="count"
          classNames="count"
          key={count}
          timeout={{ enter: 400, exit: 400 }}
        >
          <Dot>{count}</Dot>
        </CSSTransition>
      </TransitionGroup>
    </AnimationStyles>
  );
}
