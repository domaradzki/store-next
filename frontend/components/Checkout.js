import {
  CardElement,
  Elements,
  useElements,
  useStripe,
} from '@stripe/react-stripe-js';
import { loadStripe } from '@stripe/stripe-js';
import nProgress from 'nprogress';
import { useState } from 'react';
import styled from 'styled-components';
import SickButton from './styles/SickButton';

const CheckoutFormStyles = styled.form`
  box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.4);
  border: 1px solid rgba(0, 0, 0, 0.04);
  border-radius: 5px;
  padding: 1rem;
  display: grid;
  grid-gap: 1rem;
`;

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

function CheckoutForm() {
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  const stripe = useStripe();
  const elements = useElements();

  const handleSubmit = async (e) => {
    // 1.stop the form from submision and turn the loader one
    e.preventDefault();
    setLoading(true);
    console.log('We gotta do some work..');
    // 2.start the page transition
    nProgress.start();
    // 3.create the payment method via stripe (token comes back here if successful)
    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: 'card',
      card: elements.getElement(CardElement),
    });
    // TODO: stripe test cards
    console.log(paymentMethod);
    // 4. handle any errors stripe
    if (error) {
      setError(error);
    }
    // 5.send th token from step 3 to our keystone server, via a custom mutation
    // 6. change the page to view the order
    // 7. turn the loader off
    setLoading(false);
    nProgress.done();
  };
  return (
    <CheckoutFormStyles onSubmit={handleSubmit}>
      {error && <p>{error.message}</p>}
      <CardElement />
      <SickButton>Checkout now</SickButton>
    </CheckoutFormStyles>
  );
}

export default function Checkout() {
  return (
    <Elements stripe={stripeLib}>
      <CheckoutForm />
    </Elements>
  );
}
