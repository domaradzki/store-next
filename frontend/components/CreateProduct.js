import useForm from '../lib/useForm';
import Form from './styles/Form';

function CreateProduct() {
  const { inputs, handleChange, resetForm, clearForm } = useForm({
    name: 'Kosmetyk',
    description: '',
    price: 0,
    image: '',
  });

  return (
    <Form
      onSubmit={(e) => {
        e.preventDefault();
        console.log(inputs);
      }}
    >
      <fieldset>
        <label htmlFor="image">
          Image:
          <input type="file" id="image" name="image" onChange={handleChange} />
        </label>
        <label htmlFor="name">
          Name:
          <input
            type="text"
            id="name"
            name="name"
            placeholder="Name"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="description">
          Description:
          <textarea
            type="text"
            id="description"
            name="description"
            placeholder="description"
            value={inputs.description}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="price">
          Price:
          <input
            type="number"
            id="price"
            name="price"
            placeholder="Price"
            value={inputs.price}
            onChange={handleChange}
          />
        </label>
        <button type="submit">+ Add Product</button>
      </fieldset>
    </Form>
  );
}

export default CreateProduct;
