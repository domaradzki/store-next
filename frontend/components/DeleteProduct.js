import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import styled from 'styled-components';

const DELETE_PRODUCT_MUTATION = gql`
  mutation DELETE_PRODUCT_MUTATION($id: ID!) {
    deleteProduct(id: $id) {
      id
      name
    }
  }
`;

const DeleteButton = styled.button`
  cursor: pointer;
`;

function update(cache, payload) {
  console.log(payload);
  cache.evict(cache.identify(payload.data.deleteProduct));
}

export default function DeleteProduct({ id, children }) {
  const [deleteProduct, { loading }] = useMutation(DELETE_PRODUCT_MUTATION, {
    variables: { id },
    update,
  });
  return (
    <DeleteButton
      type="DeleteButton"
      disabled={loading}
      onClick={() => {
        if (confirm('Are you sure you want to delete the product')) {
          // console.log('delete');
          deleteProduct().catch((err) => alert(err));
        }
      }}
    >
      {children}
    </DeleteButton>
  );
}
