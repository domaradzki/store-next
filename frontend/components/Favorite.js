import styled from 'styled-components';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';

import {
  FavoriteBorder,
  Favorite as FavoriteFull,
} from '@styled-icons/material';
import { CURRENT_USER_QUERY, useUser } from './User';

const UPDATE_USER_FAVORITE_MUTATION = gql`
  mutation UPDATE_PRODUCT_MUTATION($id: ID!) {
    toggleFavorite(productId: $id) {
      id
    }
  }
`;

const Container = styled.div`
  position: absolute;
  cursor: pointer;
  right: 8px;
  top: 5px;
`;

const Favorite = ({ id = '' }) => {
  const user = useUser();
  const [toggleFavorite, { loading }] = useMutation(
    UPDATE_USER_FAVORITE_MUTATION,
    {
      variables: { id },
      refetchQueries: [{ query: CURRENT_USER_QUERY }],
    }
  );
  const { favorite } = user;
  const isFavorite = favorite.map((item) => item.product.id).includes(id);
  return (
    <Container disabled={loading} onClick={toggleFavorite}>
      {isFavorite ? (
        <FavoriteFull size="22" title="ulubione" color="#23263B" />
      ) : (
        <FavoriteBorder size="22" title="ulubione" color="#23263B" />
      )}
    </Container>
  );
};

export default Favorite;
