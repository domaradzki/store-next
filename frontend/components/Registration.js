import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import Form from './styles/Form';
import useForm from '../lib/useForm';
import ErrorMessage from './ErrorMessage';

export const REGISTRATION_MUTATION = gql`
  mutation REGISTRATION_MUTATION(
    $name: String!
    $email: String!
    $password: String!
  ) {
    createUser(data: { name: $name, email: $email, password: $password }) {
      id
      email
      name
    }
  }
`;

export default function Registration() {
  const { inputs, handleChange, resetForm } = useForm({
    name: '',
    email: '',
    password: '',
  });

  const [register, { error, data, loading }] = useMutation(
    REGISTRATION_MUTATION,
    {
      variables: inputs,
      // refetch the currently logged in user
      // refetchQueries: [{ query: CURRENT_USER_QUERY }],
    }
  );

  const handleSubmit = async (e) => {
    e.preventDefault();
    const res = await register().catch(console.error);
    resetForm();
  };

  return (
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Register a new Account</h2>
      <ErrorMessage error={error} />
      <fieldset>
        <label htmlFor="name">
          Name
          <input
            type="name"
            name="name"
            placeholder="Your name"
            autoComplete="name"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="Your email adress"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="password">
          Password
          <input
            type="password"
            name="password"
            placeholder="Password"
            autoComplete="password"
            value={inputs.password}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Register!</button>
      </fieldset>
    </Form>
  );
}
