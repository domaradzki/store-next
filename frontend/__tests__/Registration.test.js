import { MockedProvider } from '@apollo/react-testing';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { email } from 'casual';
import { fakeUser } from '../lib/testUtils';

import Registration, {
  REGISTRATION_MUTATION,
} from '../components/Registration';
import { CURRENT_USER_QUERY } from '../components/User';

const me = fakeUser();

const mocks = [
  // Mutation Mock
  {
    request: {
      query: REGISTRATION_MUTATION,
      variables: { name: me.name, email: me.email, password: 'git' },
    },
    result: {
      data: {
        creaAteUser: {
          __typeName: 'User',
          id: 'abc123',
          email: me.email,
          name: me.name,
        },
      },
    },
  },
  // Current user Mock
  {
    request: {
      query: CURRENT_USER_QUERY,
    },
    result: {
      data: {
        authenticatedItem: me,
      },
    },
  },
];

describe('<Registration />', () => {
  it('Renders and matches snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <Registration />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });
  it('Renders and matches snapshot', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <Registration />
      </MockedProvider>
    );
    await userEvent.type(screen.getByPlaceholderText(/name/i), me.name);
    await userEvent.type(screen.getByPlaceholderText(/email/i), me.email);
    await userEvent.type(screen.getByPlaceholderText(/password/i), 'git');
    // click submit
    await userEvent.click(screen.getByText('Register!'));
  });
});
