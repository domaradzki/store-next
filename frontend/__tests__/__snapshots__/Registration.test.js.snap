// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registration /> Renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 fDBzbp"
    method="POST"
  >
    <h2>
      Register a new Account
    </h2>
    <fieldset>
      <label
        for="name"
      >
        Name
        <input
          autocomplete="name"
          name="name"
          placeholder="Your name"
          type="name"
          value=""
        />
      </label>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="Your email adress"
          type="email"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password
        <input
          autocomplete="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Register!
      </button>
    </fieldset>
  </form>
</div>
`;
